sdcm/cluster.py

```diff
@@ -4991,8 +5012,12 @@ def node_setup(self, node, verbose=False, **kwargs):  # pylint: disable=unused-a
             self.log.info("Don't install anything because bare loaders requested")
             return
 
-        if self.params.get('client_encrypt'):
-            node.config_client_encrypt()
+        with NODE_CONFIG_SETUP_LOCK:
```

Comment:
why lock is needed?

Reply:
This is to correctly distribute individual certificates per cluster node.
The individual certificates are named the same across the nodes (the unique info such as IPs, dns names added to certificates via its SAN extension). And as the cluster nodes are created in parallel their own threads, without the lock there are situations when thread-1 creates a cert for node-1, and thread-2 creates its cert for node-2, then its time for thread-1 to copy cert-1 to node-1, but there is already cert-2.

Reply:
can we use different names or directories for each node?

Reply:
> Using different names and/or folders won't be optimal, as we would need to set different names and locations of keys/certs in server|client_encryption_details parameters in scylla.yaml, for each server.

From my understanding, we create these certs on sct-runner and then copy to server, why it cannot be created  in a 
subdir and copied to common location on server, so scylla_yaml is the same for all the nodes?

Reply:
and question, maybe to not overengineer, how much time it takes to copy&install these certs?

Reply:
> From my understanding, we create these certs on sct-runner and then copy to server, why it cannot be created in a subdir and copied to common location on server, so scylla_yaml is the same for all the nodes?

This is exactly what is performed now:
1. Before the locked section the `/etc/scylla/ssl-config` dir has only CA cert
2. Lock is acquired
4. db-to-db cert/key and db-to-client cert/key, unique for a db node, are created in the dir.
3. `/etc/scylla/ssl-config` dir is copied as a whole to a server
4. Lock is released
5. ... Repeat for other node

If I understand correctly, the concern for current approach was raised due to ```3. `ssl-config` dir is copied as a whole to a server``` step was a part of node.node_setup logic, which I also added to locked section and that is not a right approach (we would not want to lock node_setup flow).
This can be reworked so that only creating and copying operations are locked. For that, copying part should be moved out of node.node_setup, and this way node.nod_setup remains paralleled as it was.

Reply:
I think you're mixing dirs :)
`CA_CERT_FILE = get_data_dir_path('ssl_conf', 'ca.pem')` has CA cert
and we create unique per node in data dir path - that's why we cannot run in parallel.
If per node cert files were in `get_data_dir_path('ssl_conf, self.name, '<cert>') then we could run everything in parallel